server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/IndexVersionAllocationDecider.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.decider;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.routing.RecoverySource;
+import org.elasticsearch.cluster.routing.RecoverySource.SnapshotRecoverySource;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.RoutingNodes;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+
+/**
+ * An allocation decider that prevents relocation or allocation from nodes
+ * that might not be index compatible. If we relocate from a node that uses
+ * a newer index version than the node we relocate to this might cause {@link org.apache.lucene.index.IndexFormatTooNewException}
+ * on the lowest level since it might have already written segments that use a new postings format or codec that is not
+ * available on the target node.
+ */
+public class IndexVersionAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "index_version";
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (shardRouting.primary()) {
+            if (shardRouting.currentNodeId() == null) {
+                if (shardRouting.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
+                    // restoring from a snapshot - check that the node can handle the version
+                    return isSnapshotIndexVersionCompatible((SnapshotRecoverySource) shardRouting.recoverySource(), node, allocation);
+                } else {
+                    // existing or fresh primary on the node
+                    return allocation.decision(Decision.YES, NAME, "the primary shard is new or already existed on the node");
+                }
+            } else {
+                return isShardIndexVersionCompatible(shardRouting, node, allocation);
+            }
+        } else {
+            // replica - just check the index mappings version & segment updated version
+            return isShardIndexVersionCompatible(shardRouting, node, allocation);
```

Comment:
I wonder if this should check for same or newer index version of the primary vs replica nodes in addition to the check against metadata?